src/master-playlist-controller.js

```diff
@@ -0,0 +1,519 @@
+    let nextPlaylist;
+
+    // If the `error` was generated by the playlist loader, it will contain
+    // the playlist we were trying to load (but failed) and that should be
+    // blacklisted instead of the currently selected playlist which is likely
+    // out-of-date in this scenario
+    currentPlaylist = error.playlist || this.masterPlaylistLoader_.media();
+
+    // If there is no current playlist, then an error occurred while we were
+    // trying to load the master OR while we were disposing of the tech
+    if (!currentPlaylist) {
+      this.hlsHandler.error = error;
+      return this.mediaSource.endOfStream('network');
+    }
+
+    // Blacklist this playlist
+    currentPlaylist.excludeUntil = Date.now() + BLACKLIST_DURATION;
+
+    // Select a new playlist
+    nextPlaylist = this.hlsHandler.selectPlaylist();
+
+    if (nextPlaylist) {
+      videojs.log.warn('Problem encountered with the current ' +
+                       'HLS playlist. Switching to another playlist.');
+
+      return this.masterPlaylistLoader_.media(nextPlaylist);
+    }
+    videojs.log.warn('Problem encountered with the current ' +
+                     'HLS playlist. No suitable alternatives found.');
+    // We have no more playlists we can select so we must fail
+    this.hlsHandler.error = error;
+    return this.mediaSource.endOfStream('network');
+  }
+
+  pause() {
+    this.mainSegmentLoader_.pause();
+    if (this.audioPlaylistLoader_) {
+      this.audioSegmentLoader_.pause();
+    }
+  }
+
+  setCurrentTime(currentTime) {
+    let buffered = Ranges.findRange_(this.hlsHandler.tech_.buffered(), currentTime);
+
+    if (!(this.masterPlaylistLoader_ && this.masterPlaylistLoader_.media())) {
+      // return immediately if the metadata is not ready yet
+      return 0;
+    }
+
+    // it's clearly an edge-case but don't thrown an error if asked to
+    // seek within an empty playlist
+    if (!this.masterPlaylistLoader_.media().segments) {
+      return 0;
+    }
+
+    // if the seek location is already buffered, continue buffering as
+    // usual
+    if (buffered && buffered.length) {
+      return currentTime;
+    }
+
+    // cancel outstanding requests so we begin buffering at the new
+    // location
+    this.mainSegmentLoader_.abort();
+    if (this.audioPlaylistLoader_) {
+      this.audioSegmentLoader_.abort();
+    }
+
+    if (!this.hlsHandler.tech_.paused()) {
+      this.mainSegmentLoader_.load();
+      if (this.audioPlaylistLoader_) {
+        this.audioSegmentLoader_.load();
+      }
+    }
+  }
+
+  duration() {
+    if (!this.masterPlaylistLoader_) {
+      return 0;
+    }
+
+    if (this.mediaSource) {
+      return this.mediaSource.duration;
+    }
+
+    return Hls.Playlist.duration(this.masterPlaylistLoader_.media());
+  }
+
+  seekable() {
+    let media;
+    let seekable;
+
+    if (!this.masterPlaylistLoader_) {
+      return videojs.createTimeRanges();
+    }
+    media = this.masterPlaylistLoader_.media();
```

Comment:
It's legal to add variants to a live stream while it's in progress which is one way this scenario could show up. If you're running a 24/7 news channel, you might decide to add a 4k variant to your stream while it's running and there would be less video available for that variant while it filled out the live window.